### PR TITLE
Add support for 3rd party s3 providers

### PR DESCRIPTION
**Pull request by jamiechapmanbrn**
This PR adds support for S3 compatible endpoints, particularly including gcs.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ A remote object storage file system for Moodle. Intended to provide a plug-in th
 * [Currently supported object stores](#currently-supported-object-stores)
   * [Roadmap](#roadmap)
   * [Amazon S3](#amazon-s3)
+  * [Google gcs](#gcs)
```
Comment (brendanheywood) [inline README.md]: I think that needs to be #google-gcs to link properly

File: lang/en/tool_objectfs.php [hunk 2]
```diff
@@ -88,6 +88,8 @@
 $string['settings:aws:bucket_help'] = 'Amazon S3 bucket to store files in.';
 $string['settings:aws:region'] = 'region';
 $string['settings:aws:region_help'] = 'Amazon S3 API gateway region.';
+$string['settings:aws:base_url'] = 'Base URL';
+$string['settings:aws:base_url_help'] = 'Alternate url for cnames or s3 compatible endpoints';
```
Comment (brendanheywood) [inline lang/en/tool_objectfs.php]: Can you please flesh this out to make it clear that it should be left blank for normal s3

File: classes/local/store/s3/client.php [hunk 3]
```diff
@@ -72,11 +72,18 @@ public function __wakeup() {
     }
 
     public function set_client($config) {
-        $this->client = S3Client::factory(array(
-        'credentials' => array('key' => $config->s3_key, 'secret' => $config->s3_secret),
-        'region' => $config->s3_region,
-        'version' => AWS_API_VERSION
-        ));
+        # support base_url config for aws api
```
Comment (brendanheywood) [inline classes/local/store/s3/client.php]: I suspect travis doesn't like # it should be //

**Comment by brendanheywood** [comment]
thanks @jamiechapmanbrn that looks simple enough and good. I've added a few very minor comments, and can you please also fix up the little code standards bits so travis is green

**Comment by jamiechapmanbrn** [comment]
I've fixed all the review comments. The build failures look ci/network related to me.

**Comment by jamiechapmanbrn** [comment]
Updated to latest master, should be good to go.

**Comment by brendanheywood** [comment]
thanks heaps @jamiechapmanbrn

**Pull request merged**